web-app/HeimScripts/heatmap/run.r

```diff
@@ -1,92 +1,30 @@
-  grouping <- rep(dataset1color,ds1Length)
-  if(length(datasets) > 1){
-    ds2Length <- ncol(datasets[[2]]) - length(labelColumns)
-    secondSetGrouping <- rep(dataset2color,ds2Length)
-    grouping <- c(grouping,secondSetGrouping)
-  }
-  return(grouping)
-}
-
-extractMeasurements <- function(datasets){
-  measurements <- extractMeasurement(datasets[[1]])
-  if(length(datasets) > 1){
-    secondDsMeasurements <-extractMeasurement(datasets[[2]])
-    measurements <- cbind(measurements,secondDsMeasurements)
-  }
-  return(measurements)
-}
-
-extractMeasurement <- function(dataset){
-  measurements  <- subset(dataset,select=-c(Row.Label,Bio.marker)) # this will select all columns other than Row.Label,Bio.marker columns
-  measurements  <- data.matrix(measurements)
-}
-
-assignNames <- function(measurements, humanReadableRowNames){
-  rownames(measurements) <- humanReadableRowNames
-  return(measurements)
-}
-
-extractNames <- function(datasets){
-  humanReadableNames <- datasets[[1]]$Row.Label
-  #  if(length(datasets) > 1){
-  #    namesSecondDs <- datasets[[2]]$Row.Label
-  #    humanReadableNames <- c(humanReadableNames, namesSecondDs)
-  #  }
-  return(humanReadableNames)
-}
-
-
-transform <- function(measurements){
-  measurements <- log(measurements,2)
-}
-
-makeHeatmap <- function(measurements,grouping){
-  png(filename="heatmap.png", width = 800, height = 800)
-  breakpoints <- seq(from= -2,to = 10,length.out=75) # this will clamp all values below 0.1
-  breakpoints <- c(-10, breakpoints) # number of breaks must equal number of colors +1
-  heatmap.2(measurements,
-            scale = "none",
-            dendrogram = "none",
-            Rowv = NA,
-            Colv = NA,
-            density.info = "none",
-            trace = "none",
-            col=redgreen(75),
-            margins=c(12,12),
-            ColSideColors= as.character(grouping),
-            breaks = breakpoints
-  )
-  dev.off()
+   df <- loaded_variables[[1]] # SmartR does not support multiple HDD nodes yet
+   fields <- buildFields(df)
+   geneSymbols <- unique(fields["GENESYMBOL"])[,1] #[,1] in order to get a vector, otherwise we get a dataframe
+   patientIDs <-unique(fields["PATIENTID"])[,1]
+   probes <- unique(fields["PROBE"])[,1]
+   significanceValues <- unique(fields["SIGNIFICANCE"])[,1]
+   jsn <- toJSON(list("fields"=fields, "geneSymbols"=geneSymbols, 
+               "patientIDs"=patientIDs,
+               "probes"=probes,
+               "significanceValues"=significanceValues ),
+          pretty = TRUE)
+   write(jsn,file = "heatmap.json") # json file be served the same way like any other file would - get name via /status call and then /download
+}
+
+
+
+buildFields <- function(df){
+  df <- melt(df,id=names(df)[1:2])
+  SIGNIFICANCE <- rep(1.0, nrow(df))
+  raw <- df[,ncol(df)]
+  ZSCORE <- (raw - mean(raw))/ sd(raw)
+  names(df) <- c("PROBE","GENESYMBOL","PATIENTID","VALUE")
+  df["ZSCORE"] <- ZSCORE
+  df["SIGNIFICANCE"] <- SIGNIFICANCE
```

Comment:
Why calculate a significance for these z-scores? The values are not compared to another set. With the z-score you're only entering and scaling the data per gene so that the heatmap image is more neat.